### DISCUSSION TITLE
Rate limiting

### DISCUSSION THREAD

**Discussion by abuzuhri** [abuzuhri/Amazon-SP-API-CSharp#109] [Ideas]
We need to work to have better version 
https://github.com/abuzuhri/Amazon-SP-API-CSharp#usage-plans-and-rate-limits-in-the-selling-partner-api

**Comment by abuzuhri** [comment]
First version for rate limit ready for orders only 
https://github.com/abuzuhri/Amazon-SP-API-CSharp/tree/UsagePlansAndRateLimit

**Reply by abuzuhri** [reply]
@gcuadrado Can you test Rate limiting and give feedback

**Reply by gcuadrado** [reply]
Hi, ok, I will try during this weekend!

**Reply by abuzuhri** [reply]
Now rate limit ready for Library we need to test before release

**Reply by gcuadrado** [reply]
I hope to have time to test it today, I will update you

**Reply by abuzuhri** [reply]
Good news

Alpha version now ready 

https://www.nuget.org/packages/CSharpAmazonSpAPI/1.0.45.3-alpha

**Reply by asifiq2000** [reply]
> First version for rate limit ready for orders only https://github.com/abuzuhri/Amazon-SP-API-CSharp/tree/UsagePlansAndRateLimit

Hi @abuzuhri I have implemented the orders import and tracking posting features of yours FikaAmazonAPI which is running successfully
but due to high frequency of orders and trackings of some of our clients we are getting throttle limit error multiple times, which affecting
the client's productivity, so I want to implement the Rate Limit

How the Rate limit works?
Can it save me from throttle limit?
FikaAmazonAPI provides rate limit request option? Do I need to use the same FikaAmazonAPI or I need to upgrade its version?
what response we get when requests for Rate limit - as per name I think it should return the remaining request's limit. Am I right?
please answer my queries ASAP.
Thanks

**Comment by abuzuhri** [comment]
After heavy testing for Time Rate , I notes if we wait only half of mention time the app will work perfect ,Its look like if you wait the mention rate they will give you one more Burst 

@wizzardmr42
@gcuadrado
@Tilted-Software
@J-W-Chan
[wizzardmr42](https://github.com/wizzardmr42) Can you help us on test , we need to arrive to the best solution for all

**Reply by abuzuhri** [reply]
After more testing i change my mind and we should same rate mentions in API

**Reply by gcuadrado** [reply]
So no changes for now, right?

**Reply by abuzuhri** [reply]
> So no changes for now, right?

I mean for that part of divid by 2.

But now it's work perfect with new time limit you can download nuget and check it

**Reply by gcuadrado** [reply]
Oh sorry, I misunderstood you. Great! I am going to update the Nuget right now 👍

**Reply by abuzuhri** [reply]
Good , please check and give feedback

**Reply by ynaft** [reply]
I see you were testing for how much time it needs to throttle, as I did, and came to the conclusion that for requests with longer restore rates we lose a LOT of time by sleeping for the whole restore rate, I don't have a clear answer how the throttling works in the background, therefore I have no clear way to develop based on my findings, but it's a loss that we don't maximize to the possible speed.

See my exact findings here https://github.com/amzn/selling-partner-api-docs/issues/2279

**Comment by abuzuhri** [comment]
We are now fully support Rate limiting

**Comment by igoryan-k** [comment]
Hi @abuzuhri, thanks for providing the rate limiting. But it looks like I am still getting throttled on this particular function:
connection.CatalogItem.ListCatalogItemsAsync(). 
I am getting "You exceeded your quota for the requested resource." error after about 30 requests. I tried to rate limit myself to 1 request per second and still getting it even though the rate is supposed to be 5 times per second. Any ideas?  

It's a little hard to debug since the rate limit header is not exposed by the library.

**Reply by abuzuhri** [reply]
Hi @igoryan-k 

Are you sour there is no other instance running take your rate ?

**Reply by rcalv002** [reply]
Hi @abuzuhri ,

I second igoryan-k's comment. I have a singleton of this amazon service, the AmazonConnection is only initialized once, the first time the singleton is used. FYI its used by api controller to retrieve orders.

```
using AmazonIntegration.Interfaces;
using AmazonIntegration.Interfaces.SettingManagement;
using AmazonIntegration.Services.Dtos;
using FikaAmazonAPI;
using FikaAmazonAPI.AmazonSpApiSDK.Models.Orders;
using FikaAmazonAPI.Parameter.Order;
using FikaAmazonAPI.Utils;
using Volo.Abp;
using Volo.Abp.Application.Dtos;
using Volo.Abp.Threading;

namespace AmazonIntegration.Services.Amazon;

public class AmazonDomainService : IAmazonDomainService
{
    private readonly AmazonConnection _amazonConnection;

    public AmazonDomainService(IAmazonSettingsDomainService amazonSettingsDomainService)
    {
        var settingsData = new AmazonSettingsDto();

        AsyncHelper.RunSync(async () => { settingsData = await amazonSettingsDomainService.GetAsync(); });

        if (string.IsNullOrWhiteSpace(settingsData.AccessKey))
            throw new AbpException("Amazon Service Configuration data invalid",
                new ArgumentNullException(nameof(AmazonSettingsDto.AccessKey), "Value has not been configured"));

        _amazonConnection = new AmazonConnection(new AmazonCredential
        {
            AccessKey = settingsData.AccessKey,
            SecretKey = settingsData.SecretKey,
            RoleArn = settingsData.RoleArn,
            ClientId = settingsData.ClientId,
            ClientSecret = settingsData.ClientSecret,
            RefreshToken = settingsData.RefreshToken,
            MarketPlace =
                MarketPlace.GetMarketPlaceByID(settingsData
                    .MarketPlaces) //MarketPlace.GetMarketPlaceByID("ATVPDKIKX0DER") 
        });
    }

    public async Task<OrderItemList> GetOrderItems(string id)
    {
        return await _amazonConnection.Orders.GetOrderItemsAsync(id);
    }
}
```

There's also background jobs on the same process using the same singleton that may retrieve data. I routinely receive the quota exceeded message.

There's also another problem when retrieving the order items: Collection was modified; enumeration operation may not execute.
Or this An item with the same key has already been added. Key: x-amz-access-token 

Is this really built to be used as a singleton? It doesn't seem to be able to handle simultaneous requests.
If I use this as a Transient I don't get the last two errors  above but then of course quota is exceeded all the time as there's no tracking of the quota.

**Reply by abuzuhri** [reply]
Can I know which nugget version you are using?

**Reply by rcalv002** [reply]
![image](https://user-images.githubusercontent.com/25310133/167047093-e04d58f9-a01a-4aaf-aff6-14e86bcfb24b.png)

**Reply by abuzuhri** [reply]
Getting this error mean you have call get order many times before the first call finished .

I will to do small change and check with you later if that solve the problem

**Reply by igoryan-k** [reply]
In my case I used a single connection and do calls to the CatalogItem service from a single thread. I did a single call at once and waited for completion before making another call. But I was still getting the rate limiting message. 
The same function was making calls to  ProductPricing.GetItemOffers at the same rate and there was no error.

**Reply by wizzardmr42** [reply]
I don't think this approach to rate limiting will work for most use cases.  I've disabled it and I'm dealing with rate limiting by catching when there is a rate limit error and retrying after a wait period.  I have a scheduled task and a website which both use it and no way to share the limit data, plus the scheduled task loses the data when it finishes and can sometimes restart before the limits have been fully refreshed. I also found it a problem that it delays when it isn't necessary - in some cases I could make all of the calls in a burst without hitting the limit, but it still waited.

**Reply by rcalv002** [reply]
> 

@abuzuhri Were you able to check this?

**Comment by rcalv002** [comment]
Yes this sounds correct, its a one singleton and as far as i can see when you do calls to the order service it goes like this.<requeststuffhere> and so it overrides lists before stuff is done maybe? of course there are several calls coming into the api at once which would trigger individual calls out to amazon via the singleton..

Thanks for checking

**Reply by abuzuhri** [reply]
#178

**Comment by RenzoF** [comment]
Hi @abuzuhri  The rate limiter seems to be working fine for me thanks for doing that.
The only issue I have is that I use multiple "workers/threads" that handle a queue of reports and therefore the current usage is wrong.
It would be nice to abstract that so that the value could be held in memory (as it does now) but we could also use any other type of storage, like a database, thoughts?

**Reply by rcalv002** [reply]
I have the same issue! Hangfire stores the queue of transactions and the workers process these so the usage is incorrect here. Any chance we can get this implemented? Or at least option to use this way instead of standard via conf?

**Reply by RenzoF** [reply]
I think, we need to abstract this as a provider then we can have different implementations of it. One is the memory one single-threaded as currently stands and another one for the database or file system even

**Reply by wizzardmr42** [reply]
I personally find it more reliable and runs faster to disable the built-in rate limiting and always try the call, and if it hits a rate limit then wait and retry.  I don't like this approach.  @abuzuhri would you consider allowing for choosing between different rate limiting policies or allowing for custom logic to handle it so people can fine tune it if they want?

**Reply by RenzoF** [reply]
@wizzardmr42  Technically you should be able to do that from de AmazonCredential set the IsActiveLimitRate to false and handle it yourself there is also a  MaxThrottledRetryCount

**Reply by pellet8** [reply]
I have no problem using the built-in rate limits handling mechanism. We just need a way to turn off the message. It should work behind the scenes. Most people don't need to know it is working.

**Reply by TechyChap** [reply]
> Hi @abuzuhri The rate limiter seems to be working fine for me thanks for doing that. The only issue I have is that I use multiple "workers/threads" that handle a queue of reports and therefore the current usage is wrong. It would be nice to abstract that so that the value could be held in memory (as it does now) but we could also use any other type of storage, like a database, thoughts?

We also have multiple threads as our stock updates come in via rabbit messages and can have multiple consumers. So we get periodic exceptions for rate limiting when there is a sudden burst of traffic.

Perhaps an IDistributedCache could work for storing the rate limit data? The default can be in memory but many other storage providers exist.

We also have multiple instances of the application which share credentials and I'm not sure if that is also contributing to our issues.

**Reply by wizzardmr42** [reply]
@RenzoF That's what I am doing.  It feels pretty messy though - would be better if there was a more integrated solution, especially as I don't like having to deal with it on every call.

**Reply by RenzoF** [reply]
> @RenzoF That's what I am doing.  It feels pretty messy though - would be better if there was a more integrated solution, especially as I don't like having to deal with it on every call.

Agreed, what do you suggest the best approach would be? Something that works for everyone and compatible with the way that currently works

**Reply by wizzardmr42** [reply]
@RenzoF I'd design an interface that has functions on that are called before each API call and when there is a rate limit error and allow for anything that implements that interface to be used.  It will need a little thought.  Eg. I would pass in the number of retries so far on this call and the date/time it was initially requested and then have a way to return whether to retry or abort.  I'd also put in properties to control behaviour for things like throwing exceptions if it signals an abort.

**Reply by RenzoF** [reply]
We are thinking the same 👍🏼, @abuzuhri would you agree with this approach?

**Reply by abuzuhri** [reply]
It’s totally ok if the solution not affect other of course it’s ok .

i wi look to IDistributedCache and think on that

**Comment by pellet8** [comment]
How to suppress the Rate Limit message in a console application?
